zerver/views/push_notifications.py

```diff
@@ -23,23 +24,39 @@ def validate_token(token_str: str, kind: int) -> None:
         except Exception:
             raise JsonableError(_('Invalid APNS token'))
 
+def device_registration_response(token: PushDeviceToken) -> HttpResponse:
+    if settings.PUSH_NOTIFICATION_ENCRYPTION:
+        data = {'encryption_key': token.notification_encryption_key}
+        return json_success(data=data)
+
+    return json_success()
```

Comment:
This is a bad API. The request explicitly asked us to encrypt notifications... but we're deciding here *not* to encrypt the notifications, and we're returning success anyway. That is dangerous. It's reminiscent of HTTP libraries that don't check SSL/TLS certificates:
http://www.cs.utexas.edu/~shmat/shmat_ccs12.pdf

If the client explicitly asked us to encrypt notifications, and we decide we might not encrypt them after all, we should absolutely not go create a device record so that we'll start sending unencrypted notifications. We therefore should have already returned an error before this point.

I'm not sure it makes sense to have an absolute instruction like that in the request, though, given that some servers will not offer encrypted notifications. So instead of `encrypt_notifications`, we might have a parameter like `prefer_encrypted_notifications`, or `allow_encrypted_notifications`.

Reply:
Relatedly, I'm doubtful of the semantics of `settings.PUSH_NOTIFICATION_ENCRYPTION` as given in this PR. It's not described anywhere, and I am not convinced the details of it make a lot of sense.

In particular, once this is deployed with PUSH_NOTIFICATION_ENCRYPTION left false, I believe it will rapidly become impossible to change that setting to true without breaking notifications completely for some users.

We should back up and discuss these different settings -- at the server level, at the device-record level, and in the request the client makes. There should be some idea of what they each mean -- and how they interact -- and that idea should be written down somewhere. Then we can discuss that idea and make sure it makes sense, before we commit to an implementation of it.